### ISSUE TITLE
Error changes

### ISSUE THREAD

**Issue by dreson4** [ftexchange/ftx#10]
I get this error 
```
{"error":"Not logged in","success":false}
```

Whenever there's a problem with the API key, but the same error comes whenever there is any other problem for example wrong signature, timestamp, or even endpoint. 

It would be better if you could give a different error, or add error code or something for different problems so that it's easier to know where the problem comes from. If I wrongly copy the secret key, for example, there's no way I'd know that was the problem because I can't see it again once I close the dialog.

**Comment by Geesu** [comment]
@dreson4 You ever have this issue w/the order create endpoint? I get this response every time I try to use it.

**Comment by dreson4** [comment]
> @dreson4 You ever have this issue w/the order create endpoint? I get this response every time I try to use it.

Yes all endpoints gave me the same response. On my case the issue was with the signature. Once I fixed that all endpoints work just fine.

**Comment by dr497** [comment]
"Not logged in" errors usually come from a wrong signatures. FTX released an article on how to authenticate https://blog.ftx.com/blog/api-authentication/

**Comment by willp11** [comment]
Does anyone have any ideas why I am getting the not logged in error?

The code is below: 

##########################################################################
string = "https://ftx.com/api/orders"
ts = int(time.time() * 1000)
request = Request('POST', string)
prepared = request.prepare()
prepared.body = {"market": "BTC-PERP", "side": "buy", "price": 8500, "size": 0.01, "type": "limit"}

signature_payload = f'{ts}{prepared.method}{prepared.path_url}'
if prepared.body:
    signature_payload += str(prepared.body)
    
signature_payload = signature_payload.encode()
signature = hmac.new(API_secret.encode(), signature_payload, 'sha256').hexdigest()
##########################################################################

When I print signature_payload I get:

b"1593074627174POST/api/orders{'market': 'BTC-PERP', 'side': 'buy', 'price': 8500, 'size': 0.1, 'type': 'limit'}"

I read https://blog.ftx.com/blog/api-authentication/ and only changed:

signature_payload += prepared.body    ------>  signature_payload += str(prepared.body)
As get can't concatenate a dict error.

And added:

prepared.body = {"market": "BTC-PERP", "side": "buy", "price": 8500, "size": 0.01, "type": "limit"}

**Comment by dr497** [comment]
Your code should be something like:


`import urllib.parse
import time
import hmac
from requests import Request, Session

API = ""
API_secret = ""

string = "https://ftx.com/api/orders"
ts = int(time.time() * 1000)

body = {
    "market": "BTC-PERP",
    "side": "buy",
    "price": 1,
    "size": 0.0001,
    "type": "limit",
    "reduceOnly": False,
    "ioc": False,
    "postOnly": False,
    "clientId": None
}
request = Request('POST', string, json=body)
prepared = request.prepare()


signature_payload = f'{ts}{prepared.method}{prepared.path_url}'.encode()
if prepared.body:
    signature_payload += prepared.body

signature_payload = signature_payload
signature = hmac.new(API_secret.encode(),
                     signature_payload, 'sha256').hexdigest()


prepared.headers['FTX-KEY'] = API
prepared.headers['FTX-SIGN'] = signature
prepared.headers['FTX-TS'] = str(ts)
session = Session()
resp = session.send(prepared)`

**Comment by willp11** [comment]
Ok great that works now, thank you!

**Comment by mlshin** [comment]
> > @dreson4 You ever have this issue w/the order create endpoint? I get this response every time I try to use it.
> 
> Yes all endpoints gave me the same response. On my case the issue was with the signature. Once I fixed that all endpoints work just fine.

Hi
How did you fixed the problem ?

**Comment by dr497** [comment]
> issue was with the signature. Once I fixed that all endpoints work just fine.

You can follow this article to debug your signature https://blog.ftx.com/blog/api-authentication/

**Comment by mlshin** [comment]
> > issue was with the signature. Once I fixed that all endpoints work just fine.
> 
> You can follow this article to debug your signature https://blog.ftx.com/blog/api-authentication/

already  followed and seemms everythings fine . 
I am using github project and can not placeorder.

https://github.com/ftexchange/FtxApi/blob/master/FtxApi_Test/Program.cs

Timestamp is utc and format is correct.

**Comment by mlshin** [comment]
> > issue was with the signature. Once I fixed that all endpoints work just fine.
> 
> You can follow this article to debug your signature https://blog.ftx.com/blog/api-authentication/

here is my signature and seems good 

signature = "1605097628027POST/api/orders{\"market\": \"SUSHI-PERP\",\"side\": \"sell\",\"price\": 0,70,\"type\": \"market\",\"size\": 0,5,\"reduceOnly\": false}"


private string GenerateSignature(HttpMethod method, string url, string requestBody)
        {
            _nonce = GetNonce();
            var signature = $"{_nonce}{method.ToString().ToUpper()}{url}{requestBody}";
            var hash = _hashMaker.ComputeHash(Encoding.UTF8.GetBytes(signature));
            var hashStringBase64 = BitConverter.ToString(hash).Replace("-", string.Empty);
            return hashStringBase64.ToLower();
        }

**Comment by rbdm-qnt** [comment]
> Your code should be something like:
> 
> ```python
> import time
> import hmac
> from requests import Request, Session
> 
> API = ""
> API_secret = ""
> 
> string = "https://ftx.com/api/orders"
> ts = int(time.time() * 1000)
> 
> body = {
>     "market": "BTC-PERP",
>     "side": "buy",
>     "price": 1,
>     "size": 0.0001,
>     "type": "limit",
>     "reduceOnly": False,
>     "ioc": False,
>     "postOnly": False,
>     "clientId": None
> }
> 
> request = Request('POST', string, json=body)
> prepared = request.prepare()
> 
> signature_payload = f'{ts}{prepared.method}{prepared.path_url}'.encode()
> if prepared.body:
>     signature_payload += prepared.body
> 
> signature_payload = signature_payload
> signature = hmac.new(API_secret.encode(),
>                      signature_payload, 'sha256').hexdigest()
> 
> prepared.headers['FTX-KEY'] = API
> prepared.headers['FTX-SIGN'] = signature
> prepared.headers['FTX-TS'] = str(ts)
> session = Session()
> resp = session.send(prepared)
> ```

I have the same problem as @willp11. The problem is that this sample code uses Requests, and when you rewrite it using aiohttp everything works fine except for the body. This:
signature_payload += prepared.body
Has to be replaced with:
signature_payload += json.dumps(body).encode()
because you don't have a 'prepared' object. Even if the output is the exact same, for some reason it won't work. Same exact code, same exact request done with request.prepare() works perfectly. Any call that doesn't require authentication works perfectly both ways. 

Any ideas?

**Comment by anupom1998** [comment]
Please Help me...

My account show "Read-only"

Not workings my account please help me sir

**Comment by AlmasLDN** [comment]
how to i print any info?

**Comment by coding-for-money** [comment]
另外一个很重要的原因是 ：
接口文档写的接口 比如 获取用户余额的接口在api doc里写的是：/wallet/balances
但是加密的时候 不能用/wallet/balances ，而是需要用 ：/api/wallet/balances
下面是一个成功的代码例子
```java

        String host = "http://ftx.com";
        //remember ,url must add prefix "/api" 
        String url = "/api/wallet/balances";
        Long ts = System.currentTimeMillis();
        String tsStr = ts.toString();
        String apiKey = secretKeyDTO.getApiKey();
        String secretKey = secretKeyDTO.getSecretKey();
        String sign = null;
        String payload = tsStr + "GET" + url;
        sign = ThirdDigestUtil.hmacSHA256Hex(secretKey, payload);
```

**Comment by mrishan1993** [comment]
I am getting the same error for /wallet/balances. Interestingly, the same code runs for /wallet/coins and /markets. 
PLEASE HELP!! 

`const url = "https://ftx.us/api/wallet/balances"
  const path = "/api/wallet/balances"
  const timestamp = Date.now()
  const method = "GET"
  const payload = `{timestamp}{method}{url}`
  const hash = CryptoJS.HmacSHA256(payload, process.env.FTX_API_SECRET)
  // var hmac = CryptoJS.algo.HMAC.create(CryptoJS.algo.SHA256, "Secret Passphrase");
  // hmac.update(JSON.stringify(timestamp));
  // hmac.update(method);
  // hmac.update(path);
  // var hash = hmac.finalize();
  const hash2 = crypto.createHmac('sha256', process.env.FTX_API_SECRET).update(payload).digest("hex")
  console.log("API KEY ", process.env.FTX_API_KEY)
  axios({
      method: "get",
      headers: {
          "FTXUS-SIGN": CryptoJS.enc.Hex.stringify(hash),
          // "FTXUS-SIGN": hash2,
          "FTXUS-KEY": process.env.FTX_API_KEY,
          "FTXUS-TS": timestamp,
      },
      url: url
  })
  .then( (response) => {
      if (response.data.success) {
          callback(null, response.data.result)
      } else {
          // error handling here for the api 
          callback(result.data.error)
      }
      
  })
  .catch ( (e) => {
      console.log("exception in request ", e)
  })
`
